fix testing for windows

gilch opened the pull request:
Also improve `test_bin_hy_builtins` and warn that it fails from IPython.

The shell on Windows needs double quotes. This is a partial fix for #1289. It's *still* not runnning the native tests, but I wanted to push this much to make sure switching the quotes doesn't break it for Travis.

[Edit: got the native tests working on Windows too. Fixes #1289.]

Kodiologist commented:
I think we might be best off with the `shell` option of `Popen` off. Then we don't have to worry about this stuff.

gilch commented:
Like that?

gilch commented:
Hmm, nope. Travis did not like that at all.

Kodiologist commented:
With `shell` off, you need to pass `[Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen)` a list of the command and its arguments instead of a shell command as a single string.

gilch commented:
Strange, this version seems to work on Windows. It looks like the string interpretation is OS dependent, even with shell off.

I added a `shlex.split` to `run_cmd` so that any arguments in the cmd string in POSIX format will work as before. I also added an `args` kwarg, that takes a sequence to extend the list of args with, in case you have to specify them in non-POSIX format for some reason. I had to update one of the tests to do it this way, because `shlex.split` doesn't like Window's path format, but that's how `os.walk` provides it on Windows.

Kodiologist commented:
Confusingly, it looks like even when you turn `shell` off, a single string, rather than a list, will be interpreted in a platform-dependent manner. By contrast, if you pass a list, the library will [automatically](https://docs.python.org/3/library/subprocess.html#notes) convert it back into a string in the format Windows expects. It seems best to provide a list in any case, although we can still give `run_cmd` some syntactic sugar that the individual tests can use, of course.

I know this would be a bigger change than you were planning when you wrote this PR, so I'll do it if you want.

gilch commented:
> It seems best to provide a list in any case,

I'd already pushed an update to that last commit. I *am* passing a list to `Popen`.

> [...] although we can still give run_cmd some syntactic sugar that the individual tests can use, of course.

But I'm generating that list from a string passed to `run_cmd`, with `shlex.split`, which operates in a platform-independent way (i.e. POSIX sh). That's the sugar.

I could change line 231 from
```Python
output, err = run_cmd("hy2py -s -a", args=[os.path.join(dirpath, f)])
```
to
```Python
output, err = run_cmd("hy2py", args=["-s", "-a", os.path.join(dirpath, f)])
```
which puts *all* the args in a list, if you think that's clearer.

I'm not sure what else you want here. It seems to be doing everything you asked for.

Kodiologist commented:
So it does! I didn't read your latest changes closely. Sorry.

Is the change of quotation marks still necessary?

gilch commented:
> Is the change of quotation marks still necessary?

I doubt it actually. I could change it back and see.

gilch commented:
Not necessary. `tests\test_bin.py` passed on Windows when I changed it back.

I could squash the commits and omit the change to the quotes, if you want.

Kodiologist commented:
Sounds like a plan!

gilch commented:
I got pytest running the native tests on Windows.

Now this PR fixes #1289

gilch commented:
Can someone else look at this? It's making development on Windows difficult.

Kodiologist commented:
> development on Windows

There's your problem.

But seriously, Hy is mostly frozen whenever two Horsemen are inactive, and that seems to be the case right now.

kirbyfan64 commented on tests/test_bin.py:
```diff
@@ -223,18 +228,22 @@ def test_hy2py():
                 i += 1
-                output, err = run_cmd("hy2py -s -a " +
-                                      os.path.join(dirpath, f))
+                output, err = run_cmd("hy2py -s -a", args=[os.path.join(dirpath, f)])
```
Sorry for the delays, but...adding an extra argument `args=` that's really...odd (you can pass arguments through both `cmd` *and* `args`) for just *one* test case feels really wrong. A simpler change would be:

```python
output, err = run_cmd("hy2py -s -a " + shlex.quote(os.path.join(dirpath, f)))
```

(Of course, using `pipes.quote` instead on Python 2.) Then you can drop the whole `args` mess.

gilch replied:
I thought it might get used by future tests though.

gilch commented:
@kirbyfan64, how's that?
